k2/csrc/ragged_ops_inl.h

```diff
@@ -434,6 +434,158 @@ Ragged<T> Index(Ragged<T> &src, Ragged<int32_t> &indexes, bool remove_axis) {
 }
 
+
+
+namespace argmax_internal {
+template <typename T>
+struct Pair {
+  T t;
+  int32_t idx;
+};
+
+template <typename T>
+struct PairInputIterator {
+  // most of the following typedefs won't really be used; they are added because
+  // cub uses iterator_traits, which requires them.
+  typedef Pair<T> value_type;
+  typedef int32_t difference_type;
+  typedef void pointer;
+  // `reference` is a kind of dumb requirement, see this
+  // https://stackoverflow.com/questions/8110017/how-exactly-does-the-reference-typedef-behave
+  typedef Pair<T>& reference;
+  typedef std::input_iterator_tag iterator_category;
+
+  PairInputIterator(const T *t) : t_(t), offset_(0) { }
+  __device__ PairInputIterator(const T *t, int32_t offset) :
+      t_(t), offset_(offset) { }
+  __device__ PairInputIterator(const PairInputIterator &other) :
+      t_(other.t_), offset_(other.offset_) { }
+  __device__ Pair<T> operator[](int32_t idx) const {
+    return Pair<T> { t_[idx], idx + offset_ };
+  }
+  __device__ PairInputIterator operator+(int32_t offset) {
+    return PairInputIterator { t_, offset + offset_ };
+  }
+  const T *t_;
+  int32_t offset_;
+};
+
+
+template <typename T>
+struct PairOutputIteratorDeref {  // this is what you get when you dereference
+                                  // PairOutputIterator, it pretends to be a Pair<T>
+                                  // but really only stores the `idx` member.
+  __device__ PairOutputIteratorDeref(int32_t *i) : i_(i) { }
+  __device__ PairOutputIteratorDeref &operator= (const Pair<T> &p) {
+    *i_ = p.idx;
+    return *this;
+  }
+  int32_t *i_;
+};
+
+template <typename T>
+struct PairOutputIterator {  // outputs just the index of the pair.
+  // most of the following typedefs won't really be used; they are added because
+  // cub uses iterator_traits, which requires them.
+  typedef Pair<T> value_type;
+  typedef size_t difference_type;
+  typedef void pointer;
+  typedef PairOutputIteratorDeref<T> reference;
+  typedef std::output_iterator_tag iterator_category;
+
+  PairOutputIterator(int32_t *i) : i_(i), offset_(0) { }
+  __device__ PairOutputIterator(int32_t *i, size_t offset) :
+      i_(i), offset_(offset) { }
+  __device__ PairOutputIterator(const PairOutputIterator &other) :
+      i_(other.i_), offset_(other.offset_) { }
+  __device__ PairOutputIteratorDeref<T> operator[](int32_t idx) const {
+    return PairOutputIteratorDeref<T>(i_ + idx);
+  }
+  __device__ PairOutputIterator operator+(size_t offset) {
+    return PairOutputIterator { i_, offset_ + offset };
+  }
+
+  int32_t *i_;
+  size_t offset_;
+};
+
+template <typename T>
+struct PairMaxOp {
+  __device__ Pair<T> operator() (const Pair<T> &a,
+                                 const Pair<T> &b) const {
+    // NOTE: could specialize this via a union, if T == int32_t, might be
+    // marginally faster.
+    if (a.t > b.t || (a.t == b.t && a.idx > b.idx))
+      return a;
+    else return b;
+  }
+  PairMaxOp() { }
```

Comment:
I have tested and verified locally that you can remove the explicitly provided constructor and copy constructor. 
They are not needed.

Reply:
Yeah, it does't need as we did in all ops in utils.h, I have removed them locally.  Also addressed other issues of `ArgMax`, will make a PR later together.

Reply:
Removed in https://github.com/danpovey/k2/pull/4, also removed copy constructor of those Iterator and added `__forceinline__`